exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -569,28 +570,41 @@ func (r *MachinePoolReconciler) waitForMachineCreation(ctx context.Context, mach
 	return nil
 }
 
-func (r *MachinePoolReconciler) getNodeRefMap(ctx context.Context, c client.Client) (map[string]*corev1.Node, error) {
+func (r *MachinePoolReconciler) getNodeRefMap(ctx context.Context, providerIDList []string, c client.Client) (map[string]*corev1.Node, error) {
 	log := ctrl.LoggerFrom(ctx)
 	nodeRefsMap := make(map[string]*corev1.Node)
 	nodeList := corev1.NodeList{}
-	for {
-		if err := c.List(ctx, &nodeList, client.Continue(nodeList.Continue)); err != nil {
+	completeList := true
+	for _, providerID := range providerIDList {
+		if err := c.List(ctx, &nodeList, client.MatchingFields{index.NodeProviderIDField: providerID}); err != nil {
 			return nil, err
 		}
-
-		for _, node := range nodeList.Items {
-			if node.Spec.ProviderID == "" {
-				log.V(2).Info("No ProviderID detected, skipping", "providerID", node.Spec.ProviderID)
-				continue
+		if len(nodeList.Items) == 0 {
+			// If for whatever reason an index isn't registered or available, we fallback to retrieving the whole list.
+			completeList = false
+			break
+		}
+		nodeRefsMap[providerID] = &nodeList.Items[0]
+	}
+	if !completeList || len(providerIDList) == 0 {
```

Comment:
we fall back to retrieving the whole list when `len(providerIDList) == 0` because `createOrUpdateMachines` will pull the `providerID` from the `infraMachines`:
https://github.com/kubernetes-sigs/cluster-api/blob/c62c864ac9629f543bd52944e0ae64f1f9c071ad/exp/internal/controllers/machinepool_controller_phases.go#L417-L424

prior to the `providerID` list in the `machinePool.spec` getting set from the `infraConfig`: https://github.com/kubernetes-sigs/cluster-api/blob/c62c864ac9629f543bd52944e0ae64f1f9c071ad/exp/internal/controllers/machinepool_controller_phases.go#L327-L328

Reply:
unit tests such as this one: https://github.com/kubernetes-sigs/cluster-api/blob/c62c864ac9629f543bd52944e0ae64f1f9c071ad/exp/internal/controllers/machinepool_controller_phases_test.go#L1958

depend on this behavior